source/channel_bootstrap.c

```diff
@@ -23,28 +21,36 @@
 #    pragma warning(disable : 4221)
 #endif
 
-#define MAX_HOST_RESOLVER_ENTRIES 64
 #define DEFAULT_DNS_TTL 30
 
-void s_client_bootstrap_destroy_impl(struct aws_client_bootstrap *bootstrap) {
+static void s_client_bootstrap_destroy_impl(struct aws_client_bootstrap *bootstrap) {
     AWS_ASSERT(bootstrap);
     AWS_LOGF_DEBUG(AWS_LS_IO_CHANNEL_BOOTSTRAP, "id=%p: destroying", (void *)bootstrap);
     aws_client_bootstrap_shutdown_complete_fn *on_shutdown_complete = bootstrap->on_shutdown_complete;
     void *user_data = bootstrap->user_data;
+
+    aws_event_loop_group_release(bootstrap->event_loop_group);
+    aws_host_resolver_release(bootstrap->host_resolver);
+
     aws_mem_release(bootstrap->allocator, bootstrap);
 
     if (on_shutdown_complete) {
         on_shutdown_complete(user_data);
```

Comment:
I don't think either of the two cases are an issue, but I'm just skimming.  For memory checking, the assumption is that you use the aws_global_wait_thingy() call and that will not return until every elg and host resolver thread are 100% gone with no crt allocations un-released (sourced to them).  While there may be other non-crt threads (there aren't in any of the tests though), from your perspective, if you've waited then the only CRT objects left are ones you haven't released and their destruction is synchronous because all of the async things are fully gone.

Reply:
Here's the pathalogical case:
- Main thread doing `aws_global_thread_destroy_wait()`
- Nothing is currently running on the event_loop_group.
- Some object is holding the last reference to the event_loop_group.
- User releases reference to that object from a thread which is not the main thread, or an event-loop thread.
- object cleanup code calls `aws_event_loop_group_release(object->elg);` before object goes on to free its own memory.
- maybe this machine only has 1 core, or is really saturated with work
- Somehow the event_loop_group goes through its WHOLE ENTIRE WACKY CLEANUP SEQUENCE INVOLVING NEW THREADS before the `aws_event_loop_group_release()` returns. So the object still hasn't cleaned up its own memory yet.
- Main thread returns from `aws_global_thread_destroy_wait()` and goes on to check for memory leask

Reply:
Sorta makes sense but I don't think it invalidates it as a test barrier.  Doing the release from a thread that isn't an elg or the main thread that is waiting would break the reasoning though, yeah.  C++ is the only one doing the wait in its shutdown right now.  May just add calls to wait at the end of its test or something.